### ISSUE TITLE
Uniqueness validation does not extend to deleted records

### ISSUE THREAD

**Issue by dlinch** [ActsAsParanoid/acts_as_paranoid#139]
I find this issue vexing, as it's at direct odds with the documentation.

This library is mutating the default scope to include `WHERE deleted_as IS NULL`, so a uniqueness check will use this default scoping to check for unique.

So, contrary to what the docs say [here](https://github.com/ActsAsParanoid/acts_as_paranoid#validation), this library handles uniqueness checks among non-deleted records out of the box and needs some wrangling to handle uniqueness _including_ deleted records.

I was able to solve it with the `conditions` option for the uniqueness validator.

I believe the documentation needs to be updated to include the option of passing conditions to control exact scoping for uniqueness. The current validator (`validates_uniqueness_of_without_deleted`) should probably be removed as I can't see how the code would work that way.

I'm happy to open a PR with the changes, assuming I'm not missing. The below change is how I was able to get the proper rails validations to trigger on deleted records:

```
  validates  :value,
                  uniqueness: {
                    scope: :customer_id,
                    case_sensitive: false,
                    message: 'value must be unique to a customer',
                    conditions: -> { with_deleted },
                }
```

**Comment by mvz** [comment]
The documented behavior is covered in the tests, so I believe something else is going on. Do you see the same problem when the uniqueness is not scoped to the `customer_id`?

**Comment by dlinch** [comment]
I do.
My basic Model structure is:
Customer has_many tags
Tags -> acts as paranoid
  :value, :customer_id

My understanding of the documentation is that vanilla Rails validations don't account for the paranoia, so you need to add the special uniqueness validation () to be able to ignore deleted records. In my case, I want that uniqueness to extend to deleted records. I provide the ability to recover the deleted records, and I added a db constraint to prevent duplicates so I never want them to create the same value in the db for the same customer_id.

If I comment out the conditions key I mentioned in the original issue, I can create a tag, delete it, got to create a tag with the same value as the deleted one, and I get a Postgres exception.

If I comment the conditions back in, I get a proper rails validation error. I did try commenting out the scope and it didn't seem to affect this behavior.

**Comment by mvz** [comment]
I'm afraid I still can't reproduce this. When I look in the logs, the validation queries also don't include a `deleted_at` condition. You can see the default scope is removed here: https://github.com/rails/rails/blob/5-1-stable/activerecord/lib/active_record/validations/uniqueness.rb#L53.

If you can create a full minimal working app that shows this problem, or just a pull request with a failing test case, that would really help get to the bottom of this.

**Comment by dlinch** [comment]
@mvz I'm closing this issue. I just checked locally and it seems to be working. I was having a lot of issues with my paranoid logic, but I found out recently, like last week, there was a really old paranoia library in the codebase in addition to `acts_as_paranoid`. I'm fairly certain the conflicting libraries are what was exhibiting some of the strange behavior I was seeing, including the behavior I documented in this issue.

Sorry for sending you on a wild goose chase. The library in question is: https://github.com/rubysherpas/paranoia

Which uses almost the same nomenclature, but in particular is mixed in with the same call to `acts_as_paranoid`. If you do have both you see some odd behavior, which would be expected of a conflict like that.
